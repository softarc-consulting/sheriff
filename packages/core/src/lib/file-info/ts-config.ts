import { CompilerOptions } from 'typescript';

export interface TsConfig {
  compilerOptions: CompilerOptions;
  extends: string;
}
