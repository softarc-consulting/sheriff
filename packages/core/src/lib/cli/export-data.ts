import { getEntryFromCliOrConfig } from './internal/get-entry-from-cli-or-config';
import { traverseFileInfo } from '../modules/traverse-file-info';
import { calcTagsForModule } from '../tags/calc-tags-for-module';
import { ProjectInfo } from '../main/init';
import { FsPath } from '../file-info/fs-path';
import { cli } from './cli';
import getFs from '../fs/getFs';

export type ExportEntry = {
  module: string;
  tags: string[];
  imports: string[];
};

const tagsCache: Record<string, string[]> = {};

function calcOrGetTags(modulePath: FsPath, projectInfo: ProjectInfo): string[] {
  if (modulePath in tagsCache) {
    return tagsCache[modulePath];
  }

  const tags = calcTagsForModule(
    modulePath,
    projectInfo.rootDir,
    projectInfo.config.tagging,
    projectInfo.config.autoTagging,
  );

  tagsCache[modulePath] = tags;

  return tags;
}

export function createDump(entry?: string) {
  const fs = getFs();
  const projectInfo = getEntryFromCliOrConfig(entry);
  const relative = (path: FsPath) => fs.relativeTo(projectInfo.rootDir, path);

  const data: Record<string, ExportEntry> = {};

  for (const { fileInfo } of traverseFileInfo(projectInfo.fileInfo)) {
    data[relative(fileInfo.path)] = {
      module: relative(fileInfo.moduleInfo.directory) || '.',
      tags: calcOrGetTags(fileInfo.moduleInfo.directory, projectInfo),
      imports: fileInfo.imports.map((fileInfo) => relative(fileInfo.path)),
    };
  }
  return data;
}

export function exportData(...args: string[]): void {
  const data: Record<string, ExportEntry> = createDump(args[0]);
  cli.log(JSON.stringify(data, null, '  '));
}
