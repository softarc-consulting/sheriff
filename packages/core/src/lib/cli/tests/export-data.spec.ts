import { describe, it, expect } from 'vitest';
import { mockCli } from './helpers/mock-cli';
import { createProject } from '../../test/project-creator';
import { tsConfig } from '../../test/fixtures/ts-config';
import { sheriffConfig } from '../../test/project-configurator';
import { exportData } from '../export-data';
import { verifyCliWrappers } from './verify-cli-wrapper';

describe('export data', () => {
  verifyCliWrappers('export', 'src/main.ts', false);

  it('should test a simple application', () => {
    const { allLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      'sheriff.config.ts': sheriffConfig({
        tagging: {
          'src/<domain>/<type>': ['domain:<domain>', 'type:<type>'],
        },
        depRules: {},
      }),
      node_modules: {
        lodash: { 'index.js': [] },
        '@ngrx/signals': { 'index.js': [] },
        '@angular/core': { 'index.js': [] },
        '@angular/common': { 'index.js': [] },
      },
      src: {
        'main.ts': ['./holidays/feature'],
        holidays: {
          feature: {
            'index.ts': ['./holidays-container.component.ts'],
            'holidays-container.component.ts': [
              '../data',
              '../ui',
              '../model',
              '@angular/core',
              'lodash',
              '@angular/common',
            ],
          },
          data: {
            'index.ts': ['./holidays-store.ts'],
            'holidays-store.ts': ['../model', '@ngrx/signals'],
          },
          ui: {
            'index.ts': ['./holidays.component'],
            'holidays.component.ts': ['../model'],
          },
          model: { 'index.ts': ['./holiday.ts'], 'holiday.ts': [] },
        },
      },
    });

    exportData('src/main.ts');

    expect(allLogs()).toMatchSnapshot('simple-application');
  });

  it('should avoid circular dependencies', () => {
    const { allLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      src: {
        'main.ts': ['./app1.service.ts', './app2.service.ts'],
        'app1.service.ts': ['./app2.service.ts'],
        'app2.service.ts': ['./app1.service.ts'],
      },
    });

    exportData('src/main.ts');

    expect(allLogs()).toMatchSnapshot('circular-dependencies');
  });

  it('should skip not reachable files', () => {
    const { allLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      src: {
        'main.ts': ['./app1.service.ts'],
        'app1.service.ts': [],
        'app2.service.ts': [],
      },
    });

    exportData('src/main.ts');

    expect(allLogs()).toMatchSnapshot('not-reachable-files');
  });

  it('should also work with a sheriff.config.ts', () => {
    const { allLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      'sheriff.config.ts': sheriffConfig({
        tagging: { 'src/<scope>': 'scope:<scope>' },
        depRules: {},
      }),
      src: {
        'main.ts': ['./holidays', './customers'],
        holidays: {
          'index.ts': [],
        },
        customers: {
          'index.ts': [],
        },
      },
    });

    exportData('src/main.ts');

    expect(allLogs()).toMatchSnapshot('sheriff-config');
  });
});
