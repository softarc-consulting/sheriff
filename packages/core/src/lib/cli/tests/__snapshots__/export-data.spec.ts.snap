// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export data > should also work with a sheriff.config.ts > sheriff-config 1`] = `
"{
  "src/main.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/holidays/index.ts",
      "src/customers/index.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/index.ts": {
    "module": "src/holidays",
    "tags": [
      "scope:holidays"
    ],
    "imports": [],
    "externalLibraries": []
  },
  "src/customers/index.ts": {
    "module": "src/customers",
    "tags": [
      "scope:customers"
    ],
    "imports": [],
    "externalLibraries": []
  }
}"
`;

exports[`export data > should avoid circular dependencies > circular-dependencies 1`] = `
"{
  "src/main.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/app1.service.ts",
      "src/app2.service.ts"
    ],
    "externalLibraries": []
  },
  "src/app1.service.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/app2.service.ts"
    ],
    "externalLibraries": []
  },
  "src/app2.service.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/app1.service.ts"
    ],
    "externalLibraries": []
  }
}"
`;

exports[`export data > should skip not reachable files > not-reachable-files 1`] = `
"{
  "src/main.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/app1.service.ts"
    ],
    "externalLibraries": []
  },
  "src/app1.service.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [],
    "externalLibraries": []
  }
}"
`;

exports[`export data > should test a simple application > simple-application 1`] = `
"{
  "src/main.ts": {
    "module": ".",
    "tags": [
      "root"
    ],
    "imports": [
      "src/holidays/feature/index.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/feature/index.ts": {
    "module": "src/holidays/feature",
    "tags": [
      "domain:holidays",
      "type:feature"
    ],
    "imports": [
      "src/holidays/feature/holidays-container.component.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/feature/holidays-container.component.ts": {
    "module": "src/holidays/feature",
    "tags": [
      "domain:holidays",
      "type:feature"
    ],
    "imports": [
      "src/holidays/data/index.ts",
      "src/holidays/ui/index.ts",
      "src/holidays/model/index.ts"
    ],
    "externalLibraries": [
      "@angular/common",
      "@angular/core",
      "lodash"
    ]
  },
  "src/holidays/data/index.ts": {
    "module": "src/holidays/data",
    "tags": [
      "domain:holidays",
      "type:data"
    ],
    "imports": [
      "src/holidays/data/holidays-store.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/data/holidays-store.ts": {
    "module": "src/holidays/data",
    "tags": [
      "domain:holidays",
      "type:data"
    ],
    "imports": [
      "src/holidays/model/index.ts"
    ],
    "externalLibraries": [
      "@ngrx/signals"
    ]
  },
  "src/holidays/model/index.ts": {
    "module": "src/holidays/model",
    "tags": [
      "domain:holidays",
      "type:model"
    ],
    "imports": [
      "src/holidays/model/holiday.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/model/holiday.ts": {
    "module": "src/holidays/model",
    "tags": [
      "domain:holidays",
      "type:model"
    ],
    "imports": [],
    "externalLibraries": []
  },
  "src/holidays/ui/index.ts": {
    "module": "src/holidays/ui",
    "tags": [
      "domain:holidays",
      "type:ui"
    ],
    "imports": [
      "src/holidays/ui/holidays.component.ts"
    ],
    "externalLibraries": []
  },
  "src/holidays/ui/holidays.component.ts": {
    "module": "src/holidays/ui",
    "tags": [
      "domain:holidays",
      "type:ui"
    ],
    "imports": [
      "src/holidays/model/index.ts"
    ],
    "externalLibraries": []
  }
}"
`;
